File: server/src/routes/workspaces.ts
import { Hono } from 'hono';
import * as queries from '../db/queries';

export function workspaceRoutes() {
  const app = new Hono();

  app.get('/api/workspaces', (c) => {
    const workspaces = queries.listWorkspaces();
    return c.json({ workspaces });
  });

  app.get('/api/workspaces/:id', (c) => {
    const id = c.req.param('id');
    const workspace = queries.getWorkspace(id);
    if (!workspace) {
      return c.json({ error: 'Not found' }, 404);
    }
    return c.json({ workspace });
  });

  return app;
}

File: server-rust-archive/src/api/workspaces.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    routing::get,
    Json, Router,
};

use crate::db::queries;
use crate::AppState;

async fn list_workspaces(
    State(state): State<AppState>,
) -> Result<Json<serde_json::Value>, StatusCode> {
    let workspaces = queries::list_workspaces(&state.db)
        .await
        .map_err(|e| {
            tracing::error!("Failed to list workspaces: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    Ok(Json(serde_json::json!({ "workspaces": workspaces })))
}

async fn get_workspace(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, StatusCode> {
    let workspace = queries::get_workspace(&state.db, &id)
        .await
        .map_err(|e| {
            tracing::error!("Failed to get workspace: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    match workspace {
        Some(ws) => Ok(Json(serde_json::json!({ "workspace": ws }))),
        None => Err(StatusCode::NOT_FOUND),
    }
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/workspaces", get(list_workspaces))
        .route("/api/workspaces/{id}", get(get_workspace))
}
